Sign up form Rack::Attack (#42)
andycroll/andycroll.com: _posts/2019/2019-03-24-protect-your-sign-up-form-using-rack-attack.md

---
title: "Protect your sign up form with Rack::Attack"
description: "Slow down the malicious robots"
layout: article
category: ruby
image:
  base: "2019/protect-your-sign-up-form-using-rack-attack"
  alt: "Clipboard sign up"
  source: "https://unsplash.com/photos/lUShu7PHIGA"
  credit: "Eco Warrior Princess"
---

> When your application becomes popular, in may attract the attention of hackers, who’ll try and find ways to exploit the weaknesses in your site to use it for nefarious means!

Reviewer: *it may

Reviewer: No comma after 'popular'.

> They’ll nearly always explore your site manually, signing up and testing attack vectors, before attempting to automate the weaknesses they’ve established.

Reviewer: *they've discovered (right? Because you're the person who 'establishes' the weakness, in a way...)

> During an attack the hacker’s bots will typically perform “sign up with a random email then do something bad” hundreds of times a minute from a relatively small number of computers.

Reviewer: Comma after 'attack'.

Reviewer: ```suggestion
During an attack the hacker’s bots will typically sign up with a random email then do something bad, hundreds of times a minute, from a relatively small number of computers.
```

## Instead of…

> …allowing unlimited sign up attempts.

Reviewer: Nitpick point, maybe add a comma after 'attempts' instead of a full-stop since "Instead of allowing unlimited sign up attempts." isn't a proper sentence on its own.

## Use…

…`Rack::Attack` to limit the frequency of sign ups and ban the offending IP addresses.

#### `Gemfile`

```ruby
gem 'rack-attack'
```

#### `config/application.rb`

```ruby
module YourAppName
  class Application < Rails::Application
    config.middleware.use Rack::Attack
  end
end
```

#### `config/initializers/rack_attack.rb`

```ruby
class Rack::Attack
  Rack::Attack.throttle("users/sign_up", limit: 3, period: 15.minutes) do |req|
     # Using “vanilla” devise inside a User model
    req.ip if req.path == "/users" && req.post?
  end
end if Rails.env.production?
```

## But why?

The judicious use of endpoint-based request restriction can prevent your site from being an attractive target for spammers and hackers. It can also reduce the size of any successful bot attack by limiting the amount of possible signups.

> In this example they can only add three users every quarter of an hour. You _could_ likely increase the `period` as well without hitting any false positives.

Reviewer: ```suggestion
In this example, hackers can only add up to three users every quarter of an hour. You _could_ also likely increase the `period` duration without hitting any false positives.
```

To clarify, you're trying to say: you could increase the time from 15 minutes to 20/30 minutes, and it's unlikely you'd end up accidentally blocking a non-malicious user? — the sentence needed me to read it a couple of times to make sure I understood what you were saying.

> In this example, hackers can only add up to three users every quarter of an hour. You _could_ also likely increase the `period` duration to twenty or thirty minutes without accidentally blocking any legitimate users.

Reviewer: Maybe: If you think 15 minutes seems too short/brief, you _could_ likely increase... (getting rid of 'also' since there's no other thing you've discussed doing).

### Why not?

> The solution above could, _in theory_, block legitimate sign ups from potential, but it is highly unlikely that a user might incorrectly try and sign up three times in relatively quick succession.

Reviewer: ```suggestion
The solution above could, _in theory_, block legitimate sign ups, but it is highly unlikely that a user would incorrectly try and sign up three times in relatively quick succession.
```

Is 15 minutes what we'd class as 'relatively quick succession'?
